component/src/main/java/com/celements/photo/container/ImageUrl.java

```diff
@@ -0,0 +1,227 @@
+    this.action = builder.action;
+    this.space = builder.space;
+    this.name = builder.name;
+    this.filename = builder.filename;
+    this.query = builder.query;
+    this.width = builder.width;
+    this.height = builder.height;
+    this.isParsed = false;
+  }
+
+  public String getUrl() {
+    return getContext().getURLFactory().getURL(getUrlInternal(), getContext());
+  }
+
+  public String getExternalUrl() {
+    return getUrlInternal().toString();
+  }
+
+  public @NotNull Optional<String> getAction() {
+    parseUrl();
+    return Optional.fromNullable(action);
+  }
+
+  public @NotNull Optional<String> getSpace() {
+    parseUrl();
+    return Optional.fromNullable(space);
+  }
+
+  public @NotNull Optional<String> getName() {
+    parseUrl();
+    return Optional.fromNullable(name);
+  }
+
+  public @NotNull Optional<String> getFilename() {
+    parseUrl();
+    return Optional.fromNullable(filename);
+  }
+
+  public @NotNull Optional<String> getQuery() {
+    parseUrl();
+    return Optional.fromNullable(query);
+  }
+
+  public @NotNull Optional<Integer> getWidth() {
+    if (width == null) {
+      width = parseQueryDimension(CELWIDTH_FROM_URL_PATTERN);
+    }
+    if (width > 0) {
+      return Optional.of(width);
+    }
+    return Optional.absent();
+  }
+
+  public @NotNull Optional<Integer> getHeight() {
+    if (height == null) {
+      height = parseQueryDimension(CELHEIGHT_FROM_URL_PATTERN);
+    }
+    if (height > 0) {
+      return Optional.of(height);
+    }
+    return Optional.absent();
+  }
+
+  @Override
+  public String toString() {
+    return getUrl();
+  }
+
+  // TODO CELDEV-464 Make ImageUrl.parseUrl() work with prefix URL configurations
+  void parseUrl() {
+    if (!isParsed && !Strings.isNullOrEmpty(urlStr)) {
+      Matcher matcher = EXTRACT_URL_PART_PATTERN.matcher(urlStr);
+      action = getNextMatchedPart(matcher);
+      space = getNextMatchedPart(matcher);
+      name = getNextMatchedPart(matcher);
+      filename = getNextMatchedPart(matcher);
+      int queryStart = urlStr.indexOf('?');
+      if (queryStart >= 0) {
+        query = urlStr.substring(queryStart + 1);
+      } else {
+        query = "";
+      }
+      isParsed = true;
+    }
+  }
+
+  Integer parseQueryDimension(Pattern dimensionPattern) {
+    parseUrl();
+    if (!Strings.isNullOrEmpty(query)) {
+      Matcher m = dimensionPattern.matcher(query);
+      if (m.find()) {
+        String str = m.group(1);
+        try {
+          return Integer.parseInt(str);
+        } catch (NumberFormatException nfe) {
+          LOGGER.debug("Exception while parsing Integer from [{}]", str, nfe);
+        }
+      }
+    }
+    return -1;
+  }
+
+  String getNextMatchedPart(Matcher m) {
+    if (m.find()) {
+      return m.group(1);
+    }
+    return "";
+  }
+
+  URL getUrlInternal() {
+    parseUrl();
+    if (url == null) {
+      url = getContext().getURLFactory().createAttachmentURL(filename, space, name, action, query,
```

Comment:
use getters here (parseUrl can then be remove) and add defaults for action (download) and query (empty string)